subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -495,21 +535,26 @@ public boolean executeWithTask(WorkerLease parentWorkerLease, final Action<TaskI
                             @Override
                             public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
                                 ResourceLock projectLock = getProjectLock(taskInfo);
+                                if (!projectLock.tryLock()) {
+                                    return FAILED;
+                                }
+                                WorkerLease childLease = parentWorkerLease.createChild();
```

Comment:
Great point! That makes it even more efficient, because we can check the worker lease before we even check the project leases.